src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java

```diff
@@ -548,20 +576,48 @@ private void initDockerFileFile(Logger log) {
 
     private File findDockerFileFile(Logger log) {
         if (dockerFile != null) {
+            if (EnvUtil.isWindows() && !EnvUtil.isValidWindowsFileName(dockerFile)) {
+                throw new IllegalArgumentException(String.format("Invalid Windows file name %s for <dockerFile>", dockerFile));
+            }
+
             File dFile = new File(dockerFile);
-            if (dockerFileDir == null) {
+            if (dockerFileDir == null && contextDir == null) {
                 return dFile;
             } else {
-                if (dFile.isAbsolute()) {
-                    throw new IllegalArgumentException("<dockerFile> can not be absolute path if <dockerFileDir> also set.");
+                if(contextDir != null) {
+                    if(dockerFileDir != null) {
+                        log.warn("using contextDir as it has more precedence for Dockerfile");
+                    }
+
+                    if (dFile.isAbsolute()) {
+                        try {
+                            Files.copy(dFile.toPath(), Paths.get(contextDir));
```

Comment:
I don't think that the dockerfile should be copied int the `contextDir` as this is typically a source code controlled directory. Intead it should be copied into the build dir where the context is copied to for creating the tar archive to be send to the Docker daemon.

Reply:
Could you please tell me where are you creating tar archive? Is it here:

https://github.com/fabric8io/docker-maven-plugin/blob/dc82c82360afca7451b98d89f046204b506bec1e/src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java#L130-L140

Reply:
It's here where the Dockerfile is copied to its target directory: https://github.com/fabric8io/docker-maven-plugin/blob/dec36f1b75f86fa462e125172a183b7355ec460c/src/main/java/io/fabric8/maven/docker/assembly/DockerAssemblyManager.java#L196-L202

Reply:
okay, thanks

Reply:
to bo honest, I am still confused. Aren't we copying `Dockerfile` here already into build directory? Could you please tell me also where is `dockerFileDir`/`contextDir` getting copied? do we need to copy `Dockerfile` just inside build or inside the `contextDir`?

Reply:
hey, we were not done with this comment? Is it okay to merge it without addressing it?